### PR TITLE
feat(server): Add rule id to outcomes coming from event sampling

### PR DESCRIPTION
**Pull request by RaduW**
This PR adds rule ids to outcomes for events that were dropped due to event sampling.
The outcomes from these events has also changed from 3 (Invalid) to 1 (Filtered).

### CODE REVIEW

File: relay-sampling/src/lib.rs [hunk 1]
```diff
@@ -272,6 +273,16 @@ impl RuleCondition {
     }
 }
 
+/// Sampling rule Id
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
+pub struct RuleId(pub u32);
+
+impl Display for RuleId {
```
Comment (jan-auer) [inline relay-sampling/src/lib.rs:280]: To match the style used in other modules, could you please import only `std::fmt` and then qualify this as `fmt::Display`. Same for `fmt::Formatter`.

**Review by jan-auer — APPROVED** [review]

File: py/tests/test_processing.py [hunk 2]
```diff
@@ -208,7 +208,8 @@ def test_validate_sampling_configuration():
                     "op": "custom",
                     "name": "event.legacy_browser",
                     "value":["ie10"]
-                }
+                },
+                "id":1
```
Comment (jan-auer) [inline py/tests/test_processing.py:212]: This is not formatted properly, I wonder if the formatter is running on this file?

File: relay-sampling/src/lib.rs [hunk 3]
```diff
@@ -449,6 +461,9 @@ impl FieldValueProvider for TraceContext {
 pub struct SamplingConfig {
     /// The sampling rules for the project
     pub rules: Vec<SamplingRule>,
+    /// The id of the next new Rule (used as a generator for unique rule ids)
+    #[serde(default)]
+    pub next_id: Option<u32>,
```
Comment (jan-auer) [inline relay-sampling/src/lib.rs:466]: Have you considered treating this as something like the "revision" of the config object?

File: relay-server/src/utils/dynamic_sampling.rs [hunk 4]
```diff
@@ -6,40 +6,56 @@ use actix::prelude::*;
...
 use relay_general::protocol::{Event, EventId};
-use relay_sampling::{get_matching_event_rule, pseudo_random_from_uuid, rule_type_for_event};
+use relay_sampling::{
+    get_matching_event_rule, pseudo_random_from_uuid, rule_type_for_event, RuleId,
+};
 
 use crate::actors::project::{GetCachedProjectState, GetProjectState, Project, ProjectState};
 use crate::envelope::{Envelope, ItemType};
 
+/// The result of a Sampling operation
```
Comment (jan-auer) [inline relay-server/src/utils/dynamic_sampling.rs]: _nit_: Please add periods at the end of every doc string.

File: relay-server/src/utils/dynamic_sampling.rs [hunk 5]
```diff
@@ -6,40 +6,56 @@ use actix::prelude::*;
...
 
     // when we have unsupported rules disable sampling for non processing relays
     if !processing_enabled && sampling_config.has_unsupported_rules() {
-        return Some(true);
+        return SamplingResult::Keep;
     }
 
     let event_id = match event.id.0 {
-        None => return None, // if no eventID we can't really sample so keep everything
+        None => return SamplingResult::NoDecision, // if no eventID we can't really sample so keep everything
```
Comment (jan-auer) [inline relay-server/src/utils/dynamic_sampling.rs]: This comment seems out of sync. You're returning `NoDecision`, but the comment says keep. 

This is another indicator for me that the `NoDecision` enum variant may not be appropriate here. In fact, we're making the decision to explicitly keep an event, so the code should also reflect that.

Also, if we use the word "sample" in our comments, let's use the following interpretation:
- sample: keep (as in "take a sample")
- not sample: drop

**Review by jan-auer — APPROVED** [review]
More comments, still g2g.

File: py/tests/test_processing.py (hunk 2 above)
Comment (RaduW) [reply py/tests/test_processing.py:212]: hmm, I did run format on everything. The reason it is not aligned is that this is a multiline string not a dictionary.

**Pull request merged**